---
title: "gene project"
author: "akshara balachandra"
date: "5/30/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(bio3d)
library(gplots)
```

# Read the data and align.
```{r}
raw <- read.fasta('./data/all.fst')

## Align these sequences with default parameters
aln <- seqaln(raw, id=raw$id, outfile="alignment.fa", exefile = 'muscle')
```

# Generate heatmap

```{r}
alignment <- read.fasta('alignment.fa')
ident <- seqidentity(alignment)

png('img/heatmap.png', width = 1000, height = 1000)
heatmap.2(ident, key=FALSE, symkey=FALSE, density.info="none", trace="none", margins = c(21, 21), cexRow = 2, cexCol = 2)
dev.off()
```

# Generate consensus sequence
```{r}
cons.seq <- consensus(aln, cutoff = 0.5)
print(cons.seq$seq)
print(sum(cons.seq$seq != '-')/length(cons.seq$seq))
seq <- cons.seq$seq
```

# PDB search
```{r}
burmese.python <- aln$ali["Burmese_python",]
pdb.results <- blast.pdb(seq = seq, database = 'pdb')

pdb.results$hit.tbl$subjectids
# biomart r package for annotations
```
```{r}
for(i in pdb.results$hit.tbl$subjectids){
    annot <- pdb.annotate(substr(i, 1,4))
    print(paste(i, ":", annot$source, annot$compound))
}

prot.ids <- c('6DMY_B', '1VHH_A', '3K7G_B')
```

```{r}
print(pdb.annotate(prot.ids)[,c('structureId', 'compound', 'experimentalTechnique', 'resolution', 'source')])
print(pdb.results$hit.tbl[pdb.results$hit.tbl$subjectids %in% prot.ids,c('subjectids','evalue', 'identity')])

write.fasta(ids = 'Consensus Sequence 50% threshold', seqs = seq, file = './consensus_seq.fasta')
```

## FASTA sequence for 3K7G
```{r}
annot <- pdb.annotate('3K7G_B')

indian.hedgehog <- annot$sequence
write.fasta(id = 'Indian_Hedgehog_Homo_sapiens', seqs = indian.hedgehog, file = 'indian_hedgehog.fasta')

get.pdb('3K7G_B')
```


